src/Integration/Aws/DynamoDbSession/src/SessionHandler.php

```diff
@@ -238,4 +254,45 @@ private function formatKey(string $key): array
     {
         return [$this->options['hash_key'] => ['S' => $key]];
     }
+
+    /**
+     * @return array<string, AttributeValue>
+     */
+    private function readNonLocked(array $key): array
+    {
+        return $this->client->getItem([
+            'TableName' => $this->options['table_name'],
+            'Key' => $key,
+            'ConsistentRead' => $this->options['consistent_read'],
+        ])->getItem();
+    }
+
+    /**
+     * @return array<string, AttributeValue>
+     */
+    private function readLocked(array $key): array
+    {
+        $timeout = microtime(true) + $this->options['max_lock_wait_time'];
+
+        while (true) {
+            try {
+                return $this->client->updateItem([
+                    'TableName' => $this->options['table_name'],
+                    'Key' => $key,
+                    'ConsistentRead' => $this->options['consistent_read'],
+                    'Expected' => ['lock' => ['Exists' => false]],
+                    'AttributeUpdates' => ['lock' => ['Value' => ['N' => '1']]],
+                    'ReturnValues' => 'ALL_NEW',
+                ])->getAttributes();
+            } catch (ConditionalCheckFailedException $e) {
+                // If we were to exceed the timeout after sleep, let's give up immediately.
+                $sleep = rand($this->options['min_lock_retry_microtime'], $this->options['max_lock_retry_microtime']);
+                if (microtime(true) + $sleep * 1e-6 > $timeout) {
```

Comment:
This could be improved. to sleep at max "remaining time to timeout"
and throw an exception only of timeout is reached (aka when sleep < 0)